test/performance/scale_from_zero_test.go

```diff
@@ -0,0 +1,242 @@
+	return func(d *v1beta1.Deployment) (bool, error) {
+		return d.Status.ReadyReplicas == 0, nil
+	}
+}
+
+func setup(clients *test.Clients, name string) (*testContext, error) {
+	logger := logging.GetContextLogger("TestScaleFromZero (" + name + ")")
+	imagePath := test.ImagePath("helloworld")
+	names, err := createRouteAndConfig(clients, logger, imagePath, &test.Options{})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create Route and Configuration: %v", err)
+	}
+	test.CleanupOnInterrupt(func() { tearDown(clients, names, logger) }, logger)
+
+	logger.Info("Waiting for route to be marked ready.")
+	if err := test.WaitForRouteState(clients.ServingClient, names.Route, test.IsRouteReady, "RouteIsReady"); err != nil {
+		return nil, fmt.Errorf("The Route %s was not marked as Ready to serve traffic: %v", names.Route, err)
+	}
+
+	config, err := clients.ServingClient.Configs.Get(names.Config, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Configuration %s was not updated with the new revision: %v", names.Config, err)
+	}
+	deploymentName := config.Status.LatestCreatedRevisionName + "-deployment"
+
+	route, err := clients.ServingClient.Routes.Get(names.Route, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Error fetching Route %s: %v", names.Route, err)
+	}
+	domain := route.Status.Domain
+
+	return &testContext{
+		deploymentName: deploymentName,
+		logger:         logger,
+		names:          names,
+		domain:         domain,
+	}, nil
+}
+
+func runScaleFromZero(clients *test.Clients, name string) (time.Duration, error) {
+	ctx, err := setup(clients, name)
+	// setup can leak resources so we tear down even on error
+	defer tearDown(clients, ctx.names, ctx.logger)
+	if err != nil {
+		return 0, fmt.Errorf("Setup error: %v", err)
+	}
+
+	ctx.logger.Info("Waiting for deployment to scale to zero.")
+	if err = pkgTest.WaitForDeploymentState(
+		clients.KubeClient,
+		ctx.deploymentName,
+		isDeploymentScaledToZero(),
+		"DeploymentScaledToZero",
+		test.ServingNamespace,
+		2*time.Minute); err != nil {
+		return 0, fmt.Errorf("Failed waiting for deployment to scale to zero: %v", err)
+	}
+
+	start := time.Now()
+	ctx.logger.Info("Waiting for endpoint to serve request")
+	if _, err = pkgTest.WaitForEndpointState(
+		clients.KubeClient,
+		ctx.logger,
+		ctx.domain,
+		pkgTest.Retrying(pkgTest.MatchesBody(helloWorldExpectedOutput), http.StatusNotFound),
+		"HelloWorldServesText",
+		test.ServingFlags.ResolvableDomain); err != nil {
+		return 0, fmt.Errorf("The endpoint for Route %s at domain %s didn't serve the expected text \"%s\": %v", ctx.names.Route, ctx.domain, helloWorldExpectedOutput, err)
+	}
+
+	ctx.logger.Info("Request completed")
+	return time.Now().Sub(start), nil
+}
+
+func parallelScaleFromZero(t *testing.T, count int) ([]time.Duration, error) {
+	durations := make([]time.Duration, count)
+
+	clients := setupClients(t)
+	ctx := context.TODO()
+	g, ctx := errgroup.WithContext(ctx)
```

Comment:
Context shenanigans seems unneeded in this case.

```suggestion
	var g errgroup.Group
```